### PR TITLE
i32: SumSqShiftedQ31 pre-shifted sum of squares (Q31) band-energy reduction (Fixes #267)

### LINKED ISSUE #267: i32: SumSqShiftedQ31 (pre-shifted sum of squares, Q31) for band-energy reductions
## Summary

Request an `i32` primitive that computes a pre-shifted sum of squares in Q31 over an int32 slice: a wrapping-accumulated `sum += MULT32_32_Q31(x[i]<<shift, x[i]<<shift)`. This is the band-energy reduction of libopus `compute_band_energies`, and it is the last hot scalar reduction in the CELT encoder with no matching primitive here (the max-abs and dot-product reductions are already covered by `i32.MaxAbs` / `i16.DotProduct`).

## Proposed API

```go
// SumSqShiftedQ31 returns the wrapping int32 sum over a of
// MULT32_32_Q31(a[i]<<shift, a[i]<<shift), where
// MULT32_32_Q31(p, q) = int32((int64(p) * int64(q)) >> 31)
// and a[i]<<shift is an int32 left shift (wrapping). shift is in [0, 31].
func SumSqShiftedQ31(a []int32, shift int) int32
```

## Exact scalar reference (the bit-exactness target)

```go
func sumSqShiftedQ31Go(a []int32, shift int) int32 {
    var sum int32
    for _, v := range a {
        x := v << shift                       // int32, wrapping
        sum += int32((int64(x) * int64(x)) >> 31) // MULT32_32_Q31, wrapping add
    }
    return sum
}
```

Empty slice returns 0.

## Why it is SIMD-safe

Each term is an independent function of one element, and the accumulator is a wrapping int32 (two's-complement addition, associative and commutative mod 2^32). So any lane grouping and any horizontal-reduction order is bit-identical to the scalar loop for every input, exactly the property that made `DotProduct` / `MaxAbs` safe to vectorize. The only per-element intermediate is the 64-bit product before the `>> 31`, which never overflows for int32 operands, so only the final int32 cast wraps.

Widening 32x32 -> 64 multiply is available on both targets: `pmuldq` on AVX2, `sqdmull`/`smull` family on NEON, so a widening-square-plus-shift-plus-accumulate kernel maps cleanly.

## Consumer

The per-band amplitude pass in a fixed-point Opus/CELT encoder (`compute_band_energies`): for each band it derives a per-band `shift` from the band peak, sums the shifted squares, then takes a Q-domain sqrt. The primitive is the inner reduction; the shift derivation and sqrt stay in the caller. Nothing about the reduction itself is Opus-specific, which is why it belongs here rather than vendored in the consumer.

## Notes

- A variant without the pre-shift (`SumSqQ31(a)`) may be worth exposing too, but the pre-shift is what the band-energy caller needs to keep the accumulator in range, and folding it in avoids a scratch buffer.
- Happy to provide a fuzz/differential corpus (adversarial patterns around MinInt32, forced accumulator wrap, and every tail remainder) mirroring the existing `MaxAbs` / `DotProduct` suites.

### PR DESCRIPTION
**Pull request by tphakala**
Adds `i32.SumSqShiftedQ31(a []int32, shift int) int32`, the band-energy reduction of a fixed-point Opus/CELT encoder's `compute_band_energies`: the wrapping int32 sum over `a` of `MULT32_32_Q31(a[i]<<shift, a[i]<<shift)`. Closes #267.

## Semantics

`a[i]<<shift` is a wrapping int32 left shift (`SHL32`) applied before the widening multiply, so the square is of the already-truncated int32 value. The 64-bit product is at most 2^62 and never overflows the int64 intermediate; only the final int32 cast wraps (a shifted value of `MinInt32` contributes `2^62>>31 = 2^31`, wrapping to `MinInt32`). The accumulator wraps in int32 rather than saturating, so any SIMD lane grouping and horizontal reduction order is bit-identical to the sequential loop, including on overflowing inputs. `shift` must be in `[0, 31]`; out of range panics, matching `GainQ31`, since the count would otherwise diverge across backends. An empty slice returns 0. The call allocates nothing.

## Implementation

- Pure-Go reference `sumSqShiftedQ31Go` (the bit-exact source of truth), reusing the existing `gainShiftMax` / `scaleQ31Shift` constants.
- AVX2 kernel (8 int32/iter): `VPSLLD` pre-shift, `VPMULDQ` even/odd widening square, `VPSRLQ $31`, `VPBLENDD` recombine, `VPADDD` accumulate, then the `Sum`-style horizontal fold. The shift count is loaded with VEX `VMOVD` before any YMM op.
- NEON kernel (4 int32/iter): `SSHL` pre-shift, `SMULL`/`SMULL2` self-square, `SSHR #31`, `XTN`/`XTN2`, `VADD` accumulate, `VADDV` fold. The two new hand-encoded `WORD` directives are the `scaleQ31NEON` encodings with the multiplier lane set to the input, cross-checked against `arm64asm` by the asmcheck suite.
- Dispatch is a direct CPU-flag switch on all three build tags (amd64/arm64/other), one vector block on each arch; the kernels fall through to a scalar tail so they are correct at any length.

Additive change (new public function, reused constants); no existing behavior touched.

## Test plan

- `go test ./...` (native amd64 and arm64) and `go test ./asmcheck/ .`
- `go vet` for `GOARCH=amd64` and `GOARCH=arm64`
- `golangci-lint run ./...`
- Parity of each kernel against the pure-Go reference and an independent arbitrary-precision (`big.Int`) oracle across a full length sweep and all shifts; wraparound, shift-before-widen discrimination, planted-term, over-read, shift-range panic, and allocation-free assertions; a differential fuzz target (`FuzzSumSqShiftedQ31`) against the reference and the oracle.

Optional follow-ups tracked separately: #268 (a pre-existing AVX-SSE assist in `gainQ31AVX2`) and #269 (report-only micro-optimizations for these kernels).

## Summary by CodeRabbit

* **New Features**
  * Added `SumSqShiftedQ31` for calculating shifted, Q31-scaled sum-of-squares with defined wrapping `int32` behavior.
  * Supports empty inputs and shift values from 0 through 31.
  * Added optimized processing on supported AVX2 and NEON systems, with consistent results across platforms.

* **Documentation**
  * Documented the operation’s shifting, scaling, widening, and wrapping semantics.

* **Tests**
  * Added comprehensive coverage for edge cases, overflow, randomized inputs, allocation behavior, and hardware-accelerated paths.

### CODE REVIEW

File: i32/i32_amd64.s [hunk 1]
```diff
@@ -775,6 +775,82 @@ maxabs_ret_max:
...
+    JNZ  sumsq_avx2_loop8
+
+sumsq_avx2_reduce:
+    VEXTRACTI128 $1, Y7, X1
+    VPADDD X1, X7, X7            // fold 8 -> 4 int32
+    VPSHUFD $0x4E, X7, X1        // swap 64-bit halves
+    VPADDD X1, X7, X7
+    VPSHUFD $0xB1, X7, X1        // swap 32-bit within pairs
+    VPADDD X1, X7, X7
+    MOVQ X7, DX                  // EDX = vector total (low int32)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply i32/i32_amd64.s]: Fixed in 1ea37bf: the final fold extract is now VEX `VMOVD X7, DX`, so the kernel stays entirely on the VEX path through its closing `VZEROUPPER`. Only the low 32 bits (the int32 total in lane 0) are read downstream, so the result is unchanged, and go vet/build plus the AVX2 parity suite still pass.
